File: libs/language/src/lib/preprocess/processor/ReplaceInlineValuesForTypeAttributesWithImports.ts
import { BlueNode } from '../../model';
import { TransformationProcessor } from '../interfaces';
import { NodeTransformer } from '../../utils/NodeTransformer';

/**
 * Processor that replaces inline values for type attributes with imports
 */
export class ReplaceInlineValuesForTypeAttributesWithImports
  implements TransformationProcessor
{
  public static readonly MAPPINGS = 'mappings';
  private mappings: Map<string, string> = new Map();

  /**
   * Creates a new processor with the given transformation node or mappings
   * @param transformationOrMappings - The transformation node or mappings to use
   */
  constructor(transformationOrMappings: BlueNode | Map<string, string>) {
    if (transformationOrMappings instanceof BlueNode) {
      const transformation = transformationOrMappings;
      const properties = transformation.getProperties();

      if (
        properties &&
        properties[ReplaceInlineValuesForTypeAttributesWithImports.MAPPINGS]
      ) {
        const mappingsNode =
          properties[ReplaceInlineValuesForTypeAttributesWithImports.MAPPINGS];
        const mappingProperties = mappingsNode.getProperties();

        if (mappingProperties) {
          Object.entries(mappingProperties).forEach(([key, node]) => {
            const value = node.getValue();
            if (typeof value === 'string') {
              this.mappings.set(key, value);
            }
          });
        }
      }
    } else {
      this.mappings = transformationOrMappings;
    }
  }

  /**
   * Process a document node to replace inline values for type attributes with imports
   * @param document - The document to process
   * @returns The processed document
   */
  process(document: BlueNode): BlueNode {
    return NodeTransformer.transform(document, this.transformNode.bind(this));
  }

  private transformNode(node: BlueNode): BlueNode {
    const transformedNode = node.clone();
    this.transformTypeField(transformedNode, transformedNode.getType());
    this.transformTypeField(transformedNode, transformedNode.getItemType());
    this.transformTypeField(transformedNode, transformedNode.getKeyType());
    this.transformTypeField(transformedNode, transformedNode.getValueType());
    return transformedNode;
  }

  private transformTypeField(
    node: BlueNode,
    typeNode: BlueNode | undefined,
  ): void {
    if (
      typeNode &&
      typeNode.isInlineValue() &&
      typeNode.getValue() !== undefined
    ) {
      const typeValue = String(typeNode.getValue());

      if (this.mappings.has(typeValue)) {
        const blueId = this.mappings.get(typeValue);
        if (blueId) {
          const newTypeNode = new BlueNode().setBlueId(blueId);

          if (typeNode === node.getType()) {
            node.setType(newTypeNode);
          } else if (typeNode === node.getItemType()) {
            node.setItemType(newTypeNode);
          } else if (typeNode === node.getKeyType()) {
            node.setKeyType(newTypeNode);
          } else if (typeNode === node.getValueType()) {
            node.setValueType(newTypeNode);
          }
        }
      }
    }
  }
}

File: references/java-sdk/src/main/java/blue/language/preprocess/processor/ReplaceInlineValuesForTypeAttributesWithImports.java
package blue.language.preprocess.processor;

import blue.language.model.Node;
import blue.language.preprocess.TransformationProcessor;
import blue.language.utils.NodeTransformer;

import java.util.HashMap;
import java.util.Map;

public class ReplaceInlineValuesForTypeAttributesWithImports implements TransformationProcessor {

    public static final String MAPPINGS = "mappings";
    private Map<String, String> mappings = new HashMap<>();

    public ReplaceInlineValuesForTypeAttributesWithImports(Node transformation) {
        if (transformation.getProperties() != null && transformation.getProperties().containsKey(MAPPINGS)) {
            transformation.getProperties().get(MAPPINGS).getProperties().forEach((key, node) ->
                    mappings.put(key, (String) node.getValue()));
        }
    }

    public ReplaceInlineValuesForTypeAttributesWithImports(Map<String, String> mappings) {
        this.mappings = mappings;
    }

    @Override
    public Node process(Node document) {
        return NodeTransformer.transform(document, this::transformNode);
    }

    private Node transformNode(Node node) {
        Node transformedNode = node.clone();
        transformTypeField(transformedNode, transformedNode.getType());
        transformTypeField(transformedNode, transformedNode.getItemType());
        transformTypeField(transformedNode, transformedNode.getKeyType());
        transformTypeField(transformedNode, transformedNode.getValueType());
        return transformedNode;
    }

    private void transformTypeField(Node node, Node typeNode) {
        if (typeNode != null && typeNode.isInlineValue() && typeNode.getValue() != null) {
            String typeValue = typeNode.getValue().toString();
            if (mappings.containsKey(typeValue)) {
                String blueId = mappings.get(typeValue);
                Node newTypeNode = new Node().blueId(blueId);
                if (typeNode == node.getType()) {
                    node.type(newTypeNode);
                } else if (typeNode == node.getItemType()) {
                    node.itemType(newTypeNode);
                } else if (typeNode == node.getKeyType()) {
                    node.keyType(newTypeNode);
                } else if (typeNode == node.getValueType()) {
                    node.valueType(newTypeNode);
                }
            }
        }
    }
}